fodc/agent/internal/proxy/client.go

```diff
@@ -0,0 +1,700 @@
+func (c *Client) handleRegistrationStream(ctx context.Context, stream fodcv1.FODCService_RegisterAgentClient) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.stopCh:
+			return
+		default:
+		}
+
+		resp, recvErr := stream.Recv()
+		if errors.Is(recvErr, io.EOF) {
+			c.logger.Info().Msg("Registration stream closed by Proxy")
+			return
+		}
+		if recvErr != nil {
+			// Check if it's a context cancellation or deadline exceeded (expected errors during cleanup)
+			if errors.Is(recvErr, context.Canceled) || errors.Is(recvErr, context.DeadlineExceeded) {
+				c.logger.Debug().Err(recvErr).Msg("Registration stream closed")
+				return
+			}
+			if st, ok := status.FromError(recvErr); ok {
+				// Check if it's a gRPC status error with expected codes
+				code := st.Code()
+				if code == codes.Canceled || code == codes.DeadlineExceeded {
+					c.logger.Debug().Err(recvErr).Msg("Registration stream closed")
+					return
+				}
+			}
+			c.logger.Error().Err(recvErr).Msg("Error receiving from registration stream")
+			return
+		}
+
+		if resp.HeartbeatIntervalSeconds > 0 {
+			c.mu.Lock()
+			c.heartbeatInterval = time.Duration(resp.HeartbeatIntervalSeconds) * time.Second
+			c.mu.Unlock()
+
+			if c.heartbeatTicker != nil {
+				c.heartbeatTicker.Stop()
+				c.startHeartbeat(ctx)
+			}
+		}
+	}
+}
+
+// handleMetricsStream handles the metrics stream.
+func (c *Client) handleMetricsStream(ctx context.Context, stream fodcv1.FODCService_StreamMetricsClient) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.stopCh:
+			return
+		default:
+		}
+
+		resp, recvErr := stream.Recv()
+		if errors.Is(recvErr, io.EOF) {
+			c.logger.Info().Msg("Metrics stream closed by Proxy")
+			return
+		}
+		if recvErr != nil {
+			// Check if it's a context cancellation or deadline exceeded (expected errors during cleanup)
+			if errors.Is(recvErr, context.Canceled) || errors.Is(recvErr, context.DeadlineExceeded) {
+				c.logger.Debug().Err(recvErr).Msg("Metrics stream closed")
+				return
+			}
+			if st, ok := status.FromError(recvErr); ok {
+				// Check if it's a gRPC status error with expected codes
+				code := st.Code()
+				if code == codes.Canceled || code == codes.DeadlineExceeded {
+					c.logger.Debug().Err(recvErr).Msg("Metrics stream closed")
+					return
+				}
+			}
+			c.logger.Error().Err(recvErr).Msg("Error receiving from metrics stream")
+			return
+		}
+
+		filter := &MetricsRequestFilter{}
+		if resp.StartTime != nil {
+			startTime := resp.StartTime.AsTime()
+			filter.StartTime = &startTime
+		}
+		if resp.EndTime != nil {
+			endTime := resp.EndTime.AsTime()
+			filter.EndTime = &endTime
+		}
+
+		if retrieveErr := c.RetrieveAndSendMetrics(ctx, filter); retrieveErr != nil {
+			c.logger.Error().Err(retrieveErr).Msg("Failed to retrieve and send metrics")
+		}
+	}
+}
+
+// startHeartbeat starts the heartbeat ticker.
+func (c *Client) startHeartbeat(ctx context.Context) {
```

Comment:
The function is called in three different locations. Do you plan to proceed with it?